[Team-06] [iOS] SearchResultVC 로직 분리, MainVC View Layout 설정

piggyse opened the pull request:
## 구현사항
- SearchResultViewModel을 통한 SearchResult 로직 분리
- Repository 주입 -> ViewModel 주입 으로 리팩토링
- MainVC Composition Layout을 이용 Layout 설정
- Mock Repository를 이용 서버상태에 상관없이 View Test

- - -

![Simulator Screen Recording - iPhone 13 - 2022-06-03 at 13 08 14](https://user-images.githubusercontent.com/80263729/171784471-9b9c41fe-d5a3-41c6-b520-728ba1c68e2d.gif)

## 고민 및 해결
`고민`  
HearButton..
>HeartButton이 작동이 잘 되는 줄 알았으나 `MapView에서 SearchResultVC로 넘어올 때의 값을 저장하지 못하는 것`을 확인 했습니다.
  
원인은, 현재 MapView는 SearchResultVC에서 `push가 아니라 present`되고 있기때문에 SearchVC는 Deinit 되지 않고, MapView를 화면에 띄워줍니다.   하지만, HearButton에 대한 값을 업데이트 시켜주는 로직은 `ViewDidLoad때 시행되므로 업데이트가 되지 않았습니다.`
  
![스크린샷 2022-06-03 오후 1 19 58](https://user-images.githubusercontent.com/80263729/171785436-256ccffe-d79b-4854-b52e-4b6abfd1ecd9.png)

`해결`  
>ViewModel을 이용 터치된 카드의 CardIndex를 observe  
1. HouseInfoBundle 자체를 Observe한다.
-> 배열안에 IsWish값이 바뀌는 것을 확인 할 수는 있겠지만, CollectionView 전체를 Reload해야해서 비효율적이라 생각.

2. IsWish값만 따로 빼서 배열로 만들고 Observe  
-> 결국 houseInfoBundle안에 있는 값과 IsWish의 배열값을 일치시켜야하는데, 그에따른 로직이 상당히 복잡해진다고 생각

3. Touch된 CardIndex를 Observe  
-> Index에 해당하는 Cell만 reload할 수 있고, 2번방식보다는 로직이 간편하다고 생각. `채택`
 
![스크린샷 2022-06-03 오후 1 33 20](https://user-images.githubusercontent.com/80263729/171786719-38d94dd7-61a9-4da7-ae56-ec48f0fddb5b.png)

HeartButton의 문제를 해결하기 위해 ViewModel을 만들어 보았지만,  
크게보면 `VC와 Repositroy 사이에 layer가 하나더 생김으로써 더 세분화되고 로직이 분리`가 되었습니다.

## 질문 🙋‍♂️
MVVM에서 ViewModel의 값을 가져오기 위해 `Binding`을 주로 사용하는 것으로 알고 있습니다.
제가 생각하기에  Binding의 특징은 몇가지가 있습니다.
1.  값의 변화를 계속해서 추적할 수 있다.
2. CompletionHanlder와 같은 Closure를 매번 선언, 만들지 않고 bind함으로써 좀더 값을 쉽게 넘겨 받을 수 있다.
3. Observe할 수 있는 Helper Class등이 필요하다.

그렇다면, View에서 ViewModel의 값을 굳이 계속해서 추적하지 않아도 되는 것은 굳이 `bind하지 않고 그냥 주입`하면 더 편리하지 않을까요?  
혹시.. 제가 이해를 잘못한 부분이나, 리뷰어님의 의견이 궁금합니다!

piggyse committed:
[feature] Mock Protocol & TableView Header

Contents
- Mock Procotol로 서버 없이 View Test
- TableView Header 생성 및 적용

piggyse committed:
[refactor] HeartButton Delegate Refactor

Contents
- 맵뷰의 isWish값 저장 안되던 오류 수정
- HouseInfoViewModel 생성 및 HeartButton CardIndex Observe

piggyse committed:
[test] HouseInfoBundleViewModel Unittest

Contents
- fetchData test
- changeIsWish test

piggyse committed:
[feature] add TabBarController

Contents
- tabVarController 생성
- main,wish,reservation VC 임시로 생성 후 Tab에 적용

piggyse committed:
[feature] MainView 구현(CollectionView)

Contents
- MainView CollectionView(Composition Layout이용) 구현
- Custom Cell 구현 및 레이아웃 설정
- Custom DataSource 구현

lenaios commented on iOS/AirbnbApp/AirbnbApp/Model/HouseInfo.swift:
```diff
@@ -5,21 +5,22 @@
 
-struct HouseInfo: Codable {
-    let id: String = " "
+struct HouseInfo: Codable, Equatable {
```
Equatable을 채택한 이유는 뭔가요? 🧐

piggyse replied:
Test Code작성시에 제가 Mock으로 생성한 HouseInfo와 Network를 통해 가지고온 HouseInfo를 비교하기 위해서 Equatable을 채택해 보았습니다!

lenaios commented on iOS/AirbnbApp/AirbnbApp/NetworkService/NetworkManager.swift:
```diff
@@ -15,14 +15,14 @@ struct NetworkManager: NetworkManagable {
     }
     
-    func request<T: Codable>(endpoint: Endpointable, completion: @escaping ((DataResponse<T, AFError>) -> Void))  {
+    func request<T: Codable>(endpoint: Endpointable, completion: @escaping ((DataResponse<T?, AFError>) -> Void))  {
+        guard let headers = endpoint.getHeaders() else { return }
```
`private var headers: [String: String]`가 이제 옵셔널이 아니니 getHeaders()의 리턴타입도 옵셔널이 아니어도 되지 않을까요?

piggyse replied:
앗. 이부분은 제가 미처 발견하지 못했네요.
꼼꼼한 리뷰 감사합니다 😀

lenaios commented on iOS/AirbnbApp/AirbnbApp/Repository/HouseInfoRepoitoriable.swift:
```diff
@@ -0,0 +1,11 @@
+//  Created by 박진섭 on 2022/06/02.
+//
+
+protocol HouseInfoRepoitoriable {
```
s가 빠진거겠죠?

piggyse replied:
HouseInfo라는 네이밍이 조금 어색한것 같아 SearchResultRepository로 모두 변경 했습니다.

lenaios commented on iOS/AirbnbApp/AirbnbApp/ViewController/MainViewController.swift:
```diff
@@ -1,18 +1,115 @@
+        }
+    }
+    
+    private func createLayout() -> UICollectionViewCompositionalLayout? {
+        return UICollectionViewCompositionalLayout { sectionNumber, env in
+            let insetValue = 16.0
+            switch sectionNumber {
+            case 0:
+                let item = NSCollectionLayoutItem(layoutSize: NSCollectionLayoutSize(widthDimension: .fractionalWidth(1),
```
compositional layout을 사용하셨군요~ section을 0, 1, 2로 하드 코딩하지 않고 관리해주는 객체를 만들어보는 것은 어떨까요?

lenaios commented on iOS/AirbnbApp/AirbnbApp/ViewController/MainCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,68 @@
+                                                                           for: indexPath) as? MainHeaderView
+        else { return UICollectionReusableView() }
+        
+        switch indexPath.section {
+        case 1:
+            header.setLabel(text: "가까운 여행지 둘러보기")
+            return header
+        case 2:
+            header.setLabel(text: "어디에서나, 여행은 살아보는거야!")
+            return header
+        default:
+            return header
+        }
+    }
+    
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return 3
+    }
```
마찬가지로 section을 관리해주는 객체가 있다면 여기 하드코딩도 개선할 수 있어보입니다. 🙂

lenaios commented on iOS/AirbnbApp/AirbnbApp/ViewModel/SearchResultViewModel.swift:
```diff
@@ -0,0 +1,36 @@
+//  Created by 박진섭 on 2022/06/01.
+//
+
+final class SearchResultViewModel {
+    private(set) var houseInfoBundle: [HouseInfo] = []
+    private(set) var changedHeartIndex: Observable<Int> = Observable(0)
```
검색결과가 하나도 없는 경우도 있을 수 있지 않나요? 0으로 초기화 해도 되는 것인지 궁금합니다~

piggyse replied:
HeartButton의 Index는 검색결과에 따른 Cell의 갯수에 따라 하나씩 정해집니다.
만약, 검색결과가 하나도 없다면 Cell이 하나도 표시가 되지 않아서 Button자체가 보이지 않기 때문에 0으로 초기화 해도 상관없다고 생각했습니다.
하지만, 리뷰어님 말대로 검색결과가 하나도 없는데 0을 가지고 있는것은 뭔가 논리적으로 이상한 것 같다는 생각이 들어 옵셔널 값으로 수정을 진행했습니다!

lenaios reviewed (approved):
> MVVM에서 ViewModel의 값을 가져오기 위해 Binding을 주로 사용하는 것으로 알고 있습니다.
> 제가 생각하기에 Binding의 특징은 몇가지가 있습니다.
> - 값의 변화를 계속해서 추적할 수 있다.
> - CompletionHanlder와 같은 Closure를 매번 선언, 만들지 않고 bind함으로써 좀더 값을 쉽게 넘겨 받을 수 있다.
> - Observe할 수 있는 Helper Class등이 필요하다.
> 
> 그렇다면, View에서 ViewModel의 값을 굳이 계속해서 추적하지 않아도 되는 것은 굳이 bind하지 않고 그냥 주입하면 더 편리하지 않을까요?
> 혹시.. 제가 이해를 잘못한 부분이나, 리뷰어님의 의견이 궁금합니다!

뷰를 업데이트 하기 위한 뷰모델의 값이 일정해서 굳이 바인딩 하지 않아도 된다면, 뷰모델을 주입하는 것도 하나의 방법이라고 생각합니다. MVVM 이라는 큰 패턴 안에서도 객체간에 데이터를 어떻게 주고 받을 것인지 다양한 방법이 있는 것 같아요~
개인적으로는 동일한 방식을 사용해서 전체적으로 코드에 통일성을 높이는 것도 좋지만, 필요에 따라 여러 방법을 혼합해서 사용할 수도 있다고 생각합니다. 😊

piggyse commented:
> 뷰를 업데이트 하기 위한 뷰모델의 값이 일정해서 굳이 바인딩 하지 않아도 된다면, 뷰모델을 주입하는 것도 하나의 방법이라고 생각합니다. MVVM 이라는 큰 패턴 안에서도 객체간에 데이터를 어떻게 주고 받을 것인지 다양한 방법이 있는 것 같아요~ 개인적으로는 동일한 방식을 사용해서 전체적으로 코드에 통일성을 높이는 것도 좋지만, 필요에 따라 여러 방법을 혼합해서 사용할 수도 있다고 생각합니다. 😊

답변 감사합니다!
Composition Layout부분은 제가 처음써보는 거라 제가 봐도 많이 어색해 보이네요 ㅎㅎ;
조언해주신대로 수정해보겠습니다!

piggyse committed:
[Refacor] MainVC DataSource Refactory

Contents
- 각 Section별 Case문 간소화
- 각 Section별 별로 Class 생성

piggyse committed:
[feature] Add SearchBar

Contents
- SearchBar 추가

piggyse committed:
[feature] Add SearchController

piggyse committed:
[refactor] MainViewController CollectionView ReFactor

Contents
- DataSource -> DiffableDataSource
- SearchResultRepsoitory -> Respoistory
- ImageViewManager 생성

piggyse committed:
[Refactor] MainViewCell Resigtration

Contents
- MainViewRegistrator 생성 및 함수 설정

piggyse committed:
[Refactor] MainViewController 로직 분리

Contents
- MainViewregistrator 생성
- MainViewDataSourcemanager 생성

piggyse committed:
[feature] MainViewSectionViewModel

Contents
- ViewModel생성
- mockData를 통한 테스트
